src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Messaging.EventHubs.Consumer;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
+using EnsureThat;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Health.Events.Common;
+using Microsoft.Health.Events.EventCheckpointing;
+using Microsoft.Health.Events.EventConsumers.Service;
+using Microsoft.Health.Events.Telemetry;
+using Microsoft.Health.Events.Telemetry.Exceptions;
+using Microsoft.Health.Logging.Telemetry;
+
+// Description
+// - First, this class gets the number of total processors and total partitions
+// - Next, it will calculate how many partitions should be assigned to each processor
+// - Next, it will lock the partition(s) using blob leases
+// - Next, it will pass the partition id(s) into the AssignedPartitionProcessor which will run until the number of total processors changes
+
+// For a flowchart of the PartitionLockingBackgroundService see:
+// \iomt-fhir\docs\PartitionLocking.md
+
+// todos:
+// - detect if a partition has no ownership updates for some time
+// - handle cases where the ratio of processors to partitions is not divided evenly
+// - detect if acquiring partitions takes too long
+// - replace Task.Delay with something more sophisticated
+
+namespace Microsoft.Health.Events.EventHubProcessor
+{
+    public class PartitionLockingBackgroundService : BackgroundService
```

Comment:
@rogordon01 is there anyway with KEDA to pass in the total number of workers into the process?  I think if we did that model it would help eliminate a lot of complexity in the design. Then we wouldn't need to worry about determining the total number of processors.

The downside is this would likely only be applicable to our custom implementation and not apply to the generic KEDA scaler.

Reply:
imo the complexity would be reduced slightly by knowing the total number of workers but I think the majority of the complexity comes from each worker needing to know which partitions it should be working on. Passing the total number of workers in as an environment variable seems like it should be possible. But passing in the assigned partitions as an environment variable requires us to switch to kubernetes [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) which allows us to pass in different values for environment variables (the current replica set setup we have today does not allow this). Before going with the approach in this PR, I was interested in pursuing StatefulSet but did not understand the implications/conflicts with our current custom resource definition and it did not seem like an easy switch